### ISSUE TITLE
Inconsistent \parindent within size10.clo, size11.clo and size12.clo

### ISSUE THREAD

**Issue by ivankokan** [latex3/latex2e#381]
## Brief outline of the bug
Default `\parindent` lengths are inconsistently defined within `size10.clo`, `size11.clo` and `size12.clo` causing different spacing between various engines, e.g. `pdftex` and `xetex`.

size10.clo
```
\if@twocolumn
  \setlength\parindent{1em}
\else
  \setlength\parindent{15\p@}
\fi
```

size11.clo
```
\if@twocolumn
  \setlength\parindent{1em}
\else
  \setlength\parindent{17\p@}
\fi
```

size12.clo
```
\if@twocolumn
  \setlength\parindent{1em}
\else
  \setlength\parindent{1.5em}
\fi
```

Using `em`s makes the outputs of the provided MWE slightly different (note the horizontal placement of the graphics). If `10pt` or `11pt` are used instead of `12pt`, there would be no horizontal difference.

Proposal (both would bring spacing consistency between various engines):

1. Use `18\p@` instead of `1.5em` in `size12.clo`.
This would bring consistency between `.clo` files, and - more important - consistency between the `pdftex` and `xetex` output (note there is currently no consistency for `12pt` but there is for `10pt` and `11pt` for one-column layout).
2. Use `10\p@`, `11\p@` and `12\p@` respectively instead of `1em` for two-column mode.

## Minimal example showing the bug

```
\RequirePackage{latexbug}    % <- first line
\documentclass[12pt]{article}%
\usepackage{graphicx}%
\tracingall

\begin{document}%
\includegraphics{test_0}%

\the\parindent%
\end{document}%
```

## Log file (required) and possibly PDF file  

[logs.zip](https://github.com/latex3/latex2e/files/5145520/logs.zip)

**Comment by davidcarlisle** [comment]
Using em units should not make a difference between different engines if using the same fonts, although in latex xetex defaults to latin modern and pdftex defaults to computer modern. As different fonts are being used you may get different line breaking in any case, even if parindent were set using pt. (I realise that isn't the case in your test file with no text).

However the values have been as you show since the 1980s (they were essentially copied from latex 2.09 to 2e) and changing them now for pdftex would not really be possible without changing the layout of almost all existing latex documents using those options.


With pdflatex you get 

17.62482pt

and with xelatex 

18.0pt.

```
\documentclass[12pt]{article}

\begin{document}

\showthe\parindent
\end{document}
```

which was a perhaps unintended consequence of using TU encoded latin modern, however while  forcing `\parindent` to 17.62482pt would fix your test file, a normal document with text would show differences as the em spacing is different.

I think you need to set  `\parindent` explicitly for your use case or as it isn't really a paragraph at all, perhaps more naturally set it flush left with `\raggedright` or `\noindent` etc.

(personal comment, leaving open for others to comment)

**Comment by ivankokan** [comment]
> Using em units should not make a difference between different engines if using the same fonts, although in latex xetex defaults to latin modern and pdftex defaults to computer modern. As different fonts are being used you may get different line breaking in any case, even if parindent were set using pt. (I realise that isn't the case in your test file with no text).

Thanks, @davidcarlisle. I totally forgot about the LM and CM in this context! However, compiling with `\usepackage{lmodern}` still gives 17.62482pt. I guess the difference comes from the different LM font formats used by `pdftex` and `xetex`, respectively.

> I think you need to set \parindent explicitly for your use case or as it isn't really a paragraph at all, perhaps more naturally set it flush left with \raggedright or \noindent etc.

This is true. I reported the issue to discuss it generally.

> However the values have been as you show since the 1980s (they were essentially copied from latex 2.09 to 2e) and changing them now for pdftex would not really be possible without changing the layout of almost all existing latex documents using those options.

I am completely aware of it.

**Comment by FrankMittelbach** [comment]
Basically, as mentioned by David, it is effectively impossible to change the standard classes with respect to (sometimes) questionable layout decisions without causing a huge change in  millions of existing documents an people rely on them being reprocessable. We have therefore longe time ago (I think ltnews07) announced that we will not alter them if it is a question of layout improvments.

Having said that, I expect that as part of the announced large project to make LaTeX accessible and produce tagged pdf there is an opportunity to make changes to the standard classes as well without compromising backwards compatibility, simply because just like the move from 2.09 to 2e there will be eventually a distinuishing new starting declaration that identifies as document as being written for an extended LaTeX. With that in place we can at the same time introduce new updated classes that fix known layout definiciencies. With will not happen anytime soon though. So What I'm going to do is to file this issue away as longterm and move it to a "standard class improvement" project where we collect such topics to be addressed when the time is right.

**Comment by ivankokan** [comment]
@FrankMittelbach As long as the issue is understood and the change proposal (see below) is clear, I am fine with it.

```latex
%<*10pt|11pt|12pt> 
 \if@twocolumn 
 %<10pt>  \setlength\parindent{10\p@} 
 %<11pt>  \setlength\parindent{11\p@} 
 %<12pt>  \setlength\parindent{12\p@} 
 \else 
 %<10pt>  \setlength\parindent{15\p@} 
 %<11pt>  \setlength\parindent{17\p@} 
 %<12pt>  \setlength\parindent{18\p@} 
 \fi 
 %</10pt|11pt|12pt>
```

**Comment by ivankokan** [comment]
One more comment:
- `em`s are used only for `\parindent` within `size10.clo`, `size11.clo` and `size12.clo`
- there are no `ex`es within `size10.clo`, `size11.clo` and `size12.clo`.

**Comment by RuixiZhang42** [comment]
@ivankokan 
> I totally forgot about the LM and CM in this context! However, compiling with `\usepackage{lmodern}` still gives 17.62482pt. I guess the difference comes from the different LM font formats used by `pdftex` and `xetex`, respectively?

Indeed, `pdftex` will use TFM-based LM, while `xetex` will use OpenType LM. The TFM-based LM got its font metrics from the original CM, which had `1em = 18*u# = 11.75pt` for the 12pt design (see [`cmr12.mf`](http://mirrors.ctan.org/fonts/cm/mf/cmr12.mf)). This is why `1.5em` turns out to be about `17.625pt` (rounding error due to `u#:=23.5/36pt#`).

> As long as the issue is understood and the change proposal (see below) is clear, I am fine with it.
> 
> ```
> %<*10pt|11pt|12pt> 
> \if@twocolumn 
> %<10pt>  \setlength\parindent{10\p@} 
> %<11pt>  \setlength\parindent{11\p@} 
> %<12pt>  \setlength\parindent{12\p@} 
> \else 
> %<10pt>  \setlength\parindent{15\p@} 
> %<11pt>  \setlength\parindent{17\p@} % why not exactly 16.5\p@?
> %<12pt>  \setlength\parindent{18\p@} 
> \fi 
> %</10pt|11pt|12pt>
> ```

Not quite. The `11pt` in LaTeX is not 11pt, but 10.95pt (see [`ltfssbas.dtx`](https://github.com/latex3/latex2e/blob/master/base/ltfssbas.dtx)). So it should rather be
```latex
%<*10pt|11pt|12pt>
\if@twocolumn
%<10pt>  \setlength\parindent{\@xpt\p@}
%<11pt>  \setlength\parindent{\@xipt\p@}
%<12pt>  \setlength\parindent{\@xiipt\p@}
\else
%<10pt>  \setlength\parindent{15\p@}
%<11pt>  \setlength\parindent{16.425\p@}
%<12pt>  \setlength\parindent{18\p@}
\fi
%</10pt|11pt|12pt>
```

**Comment by ivankokan** [comment]
@RuixiZhang42 Thanks for sharing additional information and resources, looks great and reasonable!

To be more consistent, wouldn't `1.5\@xpt\p@` and analog be more verbose and consistent instead of using `15`, `16.425` and `18`?

**Comment by RuixiZhang42** [comment]
@ivankokan 

> To be more consistent, wouldn't `1.5\@xpt\p@` and analog be more verbose and consistent […]

The problem is that `\@xpt`, `\@xipt`, `\@xiipt` are macros. So `1.5\@xpt\p@` expands to `1.510\p@` (which is clearly *not* what we want). Even worse, `1.5\@xipt\p@` expands to `1.510.95\p@`, leading to completely wrong syntax for dimension assignment.

With LaTeX3, however, a more unified specification can be achieved using “dimension expressions” (eTeX extension):
```latex
%<10pt>  \dim_set:Nn \tex_parindent:D { \@xpt   \p@ * 3 / 2 }
%<11pt>  \dim_set:Nn \tex_parindent:D { \@xipt  \p@ * 3 / 2 }
%<12pt>  \dim_set:Nn \tex_parindent:D { \@xiipt \p@ * 3 / 2 }
```

**Comment by ivankokan** [comment]
Got it!
